### ISSUE TITLE
How to make JetBrains products sub-window float?

### ISSUE THREAD

**Issue by imkundev** [koekeishiya/yabai#194]
Hi,I want to make the JetBrains application sub window be float.I config it by following configuration.
`yabai -m rule --add title='.*Welcome*.|Checkout*.|.*Configurations|Choose*.|Import*.|.*Charges|Tip*.|Rename*.' manage=off
`
Is there any other better way?

**Comment by dominiklohmann** [comment]
> Is there any other better way?

If you're running current master, you can do this with the yet unreleased changes to signals:

Try the opposite way: float all windows, then unfloat all windows that are the only window of the application when they are created.

```sh
# regex that matches JetBrains apps
apps='^(IntelliJ IDEA|WebStorm|CLion)$'

# float all JetBrains apps
yabai -m rule --add app="${apps}" manage=off

# add a signal that floats a window created by these apps when it's the only window
yabai -m signal --add event=window_created app="${apps}" action='/path/to/some/script'
```

Now the action would need to be contained in an external, executable file (remember to `chmod +x` it), and should look something like this:

```sh
# check if there is exactly one window belonging to the process of the window that was just created
if yabai -m query --windows \
    | jq -er 'map(select(.id == env.YABAI_WINDOW_ID).pid)[0] as $pid | map(select(.pid == $pid)) | length == 1' >  /dev/null 
then
    # unfloat the window
    yabai -m window "${YABAI_WINDOW_ID}"  --toggle float
fi
```

The reason the script needs to be in an external file is that the daemon message length is limited, and this script exceeds that length by a bit. That limit will likely be removed in the future, making this a bit easier.

**Comment by kerryj89** [comment]
```
yabai -m rule --add app="PhpStorm" manage=off
yabai -m rule --add app="PhpStorm" title="~/dev/" manage=on
```

If all your projects stem from one directory like mine you can do something like that.

**Comment by imshenshen** [comment]
> > Is there any other better way?
> 
> If you're running current master, you can do this with the yet unreleased changes to signals:
> 
> Try the opposite way: float all windows, then unfloat all windows that are the only window of the application when they are created.
> 
> ```shell
> # regex that matches JetBrains apps
> apps='^(IntelliJ IDEA|WebStorm|CLion)$'
> 
> # float all JetBrains apps
> yabai -m rule --add app="${apps}" manage=off
> 
> # add a signal that floats a window created by these apps when it's the only window
> yabai -m signal --add event=window_created app="${apps}" action='/path/to/some/script'
> ```
> 
> Now the action would need to be contained in an external, executable file (remember to `chmod +x` it), and should look something like this:
> 
> ```shell
> # check if there is exactly one window belonging to the process of the window that was just created
> if yabai -m query --windows \
>     | jq -er 'map(select(.id == env.YABAI_WINDOW_ID).pid)[0] as $pid | map(select(.pid == $pid)) | length == 1' >  /dev/null 
> then
>     # unfloat the window
>     yabai -m window "${YABAI_WINDOW_ID}"  --toggle float
> fi
> ```
> 
> The reason the script needs to be in an external file is that the daemon message length is limited, and this script exceeds that length by a bit. That limit will likely be removed in the future, making this a bit easier.

如果你打开多个项目，那么他们的PID是相同的
![image](https://user-images.githubusercontent.com/19701958/63824239-ff736a80-c988-11e9-96b1-9e71fffe415e.png)
<img width="423" alt="屏幕快照 2019-08-28 上午11 42 16" src="https://user-images.githubusercontent.com/19701958/63824256-1023e080-c989-11e9-85ed-4d79adf7bc0c.png">

我觉得我们应该使用Title来判断这个文件是否存在：
```bash
~ >>> yabai -m query --windows --window 1443 | jq .title
"rds-console [~/Documents/jingdong/daasfe/rds-console] - .../client/.eslintrc.js"

~ >>> yabai -m query --windows --window 2145 | jq .title
"App.vue [/private/var/folders/3k/vs18p5xd7y145l5vq0m587640000gn/T/App.vue] - ~/Downloads/App.vue"
```
--- 

yabairc (use master branch)
```bash
# regex that matches JetBrains apps
JetBrainsApp='^(IntelliJ IDEA|WebStorm|GoLand)$'
yabai -m rule --add app="${JetBrainsApp}" manage=off
yabai -m signal --add event=window_created app="${JetBrainsApp}" action="${XDG_CONFIG_HOME}/yabai/JetBrainsApp.sh"

```
JetBrainsApp.sh
```bash
#!/bin/sh
title=$(yabai -m query --windows --window ${YABAI_WINDOW_ID} | jq .title | cut -d [ -f 2 | cut -d ] -f 1)
title="${title/\~/$HOME}"
if [ "$title" == "." ]; then
  absolute=$(pwd)
elif [ "$title" == ".." ]; then
  absolute=$(dirname "$(pwd)")
else
  absolute=$(cd "$(dirname "$title")"; pwd)/$(basename "$title")
fi
if [ -e $absolute ];then
  yabai -m window ${YABAI_WINDOW_ID} --toggle float
fi
```

**Comment by ubuntudroid** [comment]
@dominiklohmann interestingly I have to restart yabai while the IDE is running before this gets applied to the IDE windows. If I start the IDE after yabai all windows are floating... 🤔

**Comment by ubuntudroid** [comment]
Seems this broke with the latest release. All Jetbrains windows are floating now for me. :(

**Comment by cake808** [comment]
Any updates on this?
The commented scripts don't work on my machine, except for the OP one.

**Comment by kevin-cali-gcx** [comment]
Can we reopen this issue?
The proposed solutions don't seem to work :/

**Comment by cake808** [comment]
Can we reopen this issue?
The proposed solutions don't seem to work :/

**Comment by kerryj89** [comment]
@cake808 I fresh installed Catalina and experienced the issue also.

I investigated with `yabai -m query --windows` and saw that the `title` property was no longer outputting the path within it, which is what my solution hooked onto.

One thing I love about JetBrains is that they have a setting to revert to the original functionality.

Go to `PHPStorm > Preferences > Appearance & behavior > Appearance > UI Options > Always show full path in window header`.  You might not need to do the following but I did it anyway and it started working like before: Restart PHPStorm and Yabai.

**Comment by shepherdjerred** [comment]
I was able to achieve it with this for IntelliJ:
```
yabai -m rule --add app="IntelliJ IDEA" manage=off
yabai -m rule --add app="IntelliJ IDEA" title=".*\[(.*)\].*" manage=on
```

And, as @kerryj89 said:
> Go to IntelliJ IDEA > Preferences > Appearance & behavior > Appearance > UI Options > Always show full path in window header

**Comment by ubuntudroid** [comment]
@shepherdjerred I've slightly modified your rules to match gradle files in Android Studio (their title doesn't seem to be affected by the "Always show full path in window header"):

yabai -m rule --add app="^Android Studio" manage=off
yabai -m rule --add app="^Android Studio" title="(.*\[(.*)\].*)|(.*\(\:.*\).*)" manage=on

**Comment by carlosflorencio** [comment]
I just want to add that is useful to have the config `window_topmost` off to prevent having some IDE action popups being inacessible.

`yabai -m config window_topmost off` Brings these type of popups to the front of the application:

![image](https://user-images.githubusercontent.com/1500881/123004767-ff89f900-d3ac-11eb-833f-a0ec1dc8b62e.png)

**Comment by shepherdjerred** [comment]
@ubuntudroid I think your rules work even better than mine. Thanks for contributing!

**Comment by carlosuno** [comment]
> @shepherdjerred I've slightly modified your rules to match gradle files in Android Studio (their title doesn't seem to be affected by the "Always show full path in window header"):
> 
> ```
> yabai -m rule --add app="^Android Studio" manage=off
> yabai -m rule --add app="^Android Studio" title="(.*\[(.*)\].*)|(.*\(\:.*\).*)" manage=on
> ```

Thank you @ubuntudroid for your solution. 
You can also do it on one line by reversing the search on the title [!]
`yabai -m rule --add app="^Android Studio" title!="(.*\[(.*)\].*)|(.*\(\:.*\).*)" manage=off`

**Comment by ubuntudroid** [comment]
@carlosuno Neat, didn't know about that reverse operator! Many thanks! 🙏

**Comment by kodokoto** [comment]
I am trying to do the same for Ableton and 3rd party plugins, unfortunately Ableton does not have an option to show the full path in the title, therefore the latest solution does not work.

I did find that there is an option for Ableton called `-ShowFullVersionInTitle` from [this list](https://www.reddit.com/r/ableton/comments/g377ms/full_list_of_optionstxt_commands_for_ableton_live/fnr3n8l?utm_source=share&utm_medium=web2x&context=3) however i cant seem to find a way to get it working.

Here is an example of what I get when I use `yabai -m query --windows`:
```
{
	"id":981,
	"pid":13984,
	"app":"Live",
	"title":"Untitled",
	"frame":{
		"x":20.0000,
		"y":20.0000,
		"w":1890.0000,
		"h":1560.0000
	},
	"level":0,
	"role":"AXUnknown",
	"subrole":"AXStandardWindow",
	"movable":1,
	"resizable":1,
	"display":1,
	"space":2,
	"visible":0,
	"focused":0,
	"split":"vertical",
	"floating":0,
	"sticky":0,
	"minimized":0,
	"topmost":0,
	"opacity":1.0000,
	"shadow":1,
	"border":0,
	"stack-index":0,
	"zoom-parent":0,
	"zoom-fullscreen":0,
	"native-fullscreen":0
},{
	"id":1010,
	"pid":13984,
	"app":"Live",
	"title":"FF Pro-L 2/Master",
	"frame":{
		"x":1930.0000,
		"y":20.0000,
		"w":950.0000,
		"h":589.0000
	},
	"level":3,
	"role":"AXWindow",
	"subrole":"AXFloatingWindow",
	"movable":1,
	"resizable":0,
	"display":1,
	"space":2,
	"visible":0,
	"focused":0,
	"split":"vertical",
	"floating":0,
	"sticky":0,
	"minimized":0,
	"topmost":1,
	"opacity":1.0000,
	"shadow":1,
	"border":0,
	"stack-index":0,
	"zoom-parent":0,
	"zoom-fullscreen":0,
	"native-fullscreen":0
}
```

Unfortunately I cant seem to do something like `yabai -m rule --add app="^Live$" role="AXUnknown" manage=on`.

I also tried the initial solution used in this thread but that does not seem to work anymore.

Please let me know if this is the wrong place to be asking for help.

**Comment by hld0** [comment]
Unfortunately yabai doesn't manage my IntelliJ after using the solution provided by @carlosuno - does anyone have a configuration specifically for IntelliJ IDEA?

**Comment by Frankcs96** [comment]
my intellij is also not managed by yabai, only if I restart with the IDE open :(

**Comment by iantanwx** [comment]
> Unfortunately yabai doesn't manage my IntelliJ after using the solution provided by @carlosuno - does anyone have a configuration specifically for IntelliJ IDEA?

Works for me:
```
yabai -m rule --add app="^IntelliJ IDEA$" manage=off
yabai -m rule --add app="^IntelliJ IDEA$" title="( – )" space=7 manage=on
```

**Comment by kunpehx** [comment]
> > Unfortunately yabai doesn't manage my IntelliJ after using the solution provided by @carlosuno - does anyone have a configuration specifically for IntelliJ IDEA?
> 
> Works for me:
> 
> ```
> yabai -m rule --add app="^IntelliJ IDEA$" manage=off
> yabai -m rule --add app="^IntelliJ IDEA$" title="( – )" space=7 manage=on
> ```

Great!It's working again,🤪。
```sh
yabai -m rule --add app="^IntelliJ IDEA-EAP$" manage=off
yabai -m rule --add app="^IntelliJ IDEA-EAP$" title="( – )"  manage=on
```

**Comment by proskehy** [comment]
> my intellij is also not managed by yabai, only if I restart with the IDE open :(

I'm still experiencing this issue. 
Neither this:
```
yabai -m rule --add app="^IntelliJ IDEA$" manage=off
yabai -m rule --add app="^IntelliJ IDEA$" title="( – )" space=7 manage=on
```
nor this
```
yabai -m rule --add app="IntelliJ IDEA" manage=off
yabai -m rule --add app="IntelliJ IDEA" title=".*\[(.*)\].*" manage=on
```
seems to work.

Am I doing something wrong or is the mentioned issue still present?

**Comment by persiyanov** [comment]
> > my intellij is also not managed by yabai, only if I restart with the IDE open :(
> 
> I'm still experiencing this issue. Neither this:
> 
> ```
> yabai -m rule --add app="^IntelliJ IDEA$" manage=off
> yabai -m rule --add app="^IntelliJ IDEA$" title="( – )" manage=on
> ```
> 
> nor this
> 
> ```
> yabai -m rule --add app="IntelliJ IDEA" manage=off
> yabai -m rule --add app="IntelliJ IDEA" title=".*\[(.*)\].*" manage=on
> ```
> 
> or this
> 
> ```
> yabai -m rule --add app="^IntelliJ IDEA$" title!=".*\[(.*)\].*" manage=off
> ```
> 
> seems to work.
> 
> Am I doing something wrong or is the mentioned issue still present?

Same. By applying these changes (I'm using `app="PyCharm"`), all IDE windows become floating. PyCharm 2022.3.1, yabai-v5.0.2, macos Monterey 12.6.3.

**Comment by Kurounin** [comment]
> > > my intellij is also not managed by yabai, only if I restart with the IDE open :(
> > 
> > 
> > I'm still experiencing this issue. Neither this:
> > ```
> > yabai -m rule --add app="^IntelliJ IDEA$" manage=off
> > yabai -m rule --add app="^IntelliJ IDEA$" title="( – )" manage=on
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > nor this
> > ```
> > yabai -m rule --add app="IntelliJ IDEA" manage=off
> > yabai -m rule --add app="IntelliJ IDEA" title=".*\[(.*)\].*" manage=on
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or this
> > ```
> > yabai -m rule --add app="^IntelliJ IDEA$" title!=".*\[(.*)\].*" manage=off
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > seems to work.
> > Am I doing something wrong or is the mentioned issue still present?
> 
> Same. By applying these changes (I'm using `app="PyCharm"`), all IDE windows become floating. PyCharm 2022.3.1, yabai-v5.0.2, macos Monterey 12.6.3.

I'm using the following config and it works on MacOS Ventura 13.2.1 and yabai-v5.0.2:
```
# make JetBrains products popup windows float
apps='^(IntelliJ IDEA|WebStorm|GoLand|PyCharm)$'
yabai -m rule --add app="JetBrains Toolbox" manage=off
yabai -m rule --add app="${apps}" manage=off
yabai -m rule --add app="${apps}" title="( – )" manage=on
```

**Comment by ubuntudroid** [comment]
The following `skhd` shortcut might help as a workaround:

```
# float / unfloat window and center on screen
shift + alt - t : yabai -m window --toggle float;\
          yabai -m window --grid 4:4:1:1:2:2
```

Essentially, whenever there is a window that behaves oddly (e.g., float even if it shouldn't), select that window and invoke that shortcut once to toggle floating for it. Works like a charm for me every time.

The center part (second line) of the command is optional. Feel free to remove it if you don't need it.

Disclosure: I don't remember whether this command was part of `skhd`'s default configuration, or if I copied it somewhere else. One thing I'm sure is, that it's certainly not mine. 😅

**Comment by Apolsus** [comment]
@Kurounin 
Your code works well.

```
# make JetBrains products popup windows float
apps='^(IntelliJ IDEA|WebStorm|GoLand|PyCharm)$'
yabai -m rule --add app="JetBrains Toolbox" manage=off
yabai -m rule --add app="${apps}" manage=off
yabai -m rule --add app="${apps}" title="( – )" manage=on
```
However, when the project was first opened, this seemed not work.

![CleanShot 2023-04-01 at 05 20 25](https://user-images.githubusercontent.com/59809602/229234037-7776963b-d42a-4212-8898-fd3aaa8d24a1.gif)

**Comment by Apolsus** [comment]
https://github.com/koekeishiya/yabai/issues/1687#issuecomment-1492866141

I have solved this in #1687

**Comment by gldtn** [comment]
> > Unfortunately yabai doesn't manage my IntelliJ after using the solution provided by @carlosuno - does anyone have a configuration specifically for IntelliJ IDEA?
> 
> Works for me:
> 
> ```
> yabai -m rule --add app="^IntelliJ IDEA$" manage=off
> yabai -m rule --add app="^IntelliJ IDEA$" title="( – )" space=7 manage=on
> ```

This works for me in PhpStorm, my only issue now is with the popups are being very annoying where Yabai thinks it's a new window for every small popup that appears in the IDE.

![phpstorm-yabai](https://github.com/koekeishiya/yabai/assets/981383/c6744f53-df92-42b6-95bf-88b5e191eda8)

Has anyone else encountered this? Any idea on how I can try to manipulate these popups from being detected as a new window?

Thanks!
